### PR TITLE
Add documentations for prior module

### PR DESCRIPTION
**Pull request by zhengp0**
This is a PR for adding docstring to the `prior` module.
Any comment is appreciated! Thanks very much!

### CODE REVIEW

File: src/regmod/prior.py [hunk 1]
```diff
@@ -13,9 +13,48 @@
 
 @dataclass
 class Prior:
+    """Prior information for the variables, it is used to construct the
+    liklihood and solve the optimization problem.
```
Comment (kels271828) [inline src/regmod/prior.py]: Typo: likelihood

File: src/regmod/prior.py [hunk 2]
```diff
@@ -13,9 +13,48 @@
...
+    Parameters
+    ----------
+    size : Optional[int], optional
+        Size of variable. Default is `None`. When it is `None`, size is inferred
+        from the vector information of the prior.
+
+    Attributes
+    ----------
+    size : int
+        Size of variable
```
Comment (kels271828) [inline src/regmod/prior.py]: Should end with a period.

File: src/regmod/prior.py [hunk 3]
```diff
@@ -13,9 +13,48 @@
...
+
     size: int = None
 
     def process_size(self, vecs: List[Any]):
+        """Infer and validate size from given vector information.
+
+        Parameters
+        ----------
+        vecs : List[Any]
+            Vector infromation of the prior. For Gaussian prior it will be mean
```
Comment (kels271828) [inline src/regmod/prior.py]: Typo: information

File: src/regmod/prior.py [hunk 4]
```diff
@@ -27,30 +66,137 @@ def process_size(self, vecs: List[Any]):
...
-        self.sd = default_vec_factory(self.sd, self.size, np.inf, vec_name='sd')
-        assert all(self.sd > 0.0), "Standard deviation must be all positive."
+        if np.isscalar(self.mean):
+            self.mean = np.repeat(self.mean, self.size)
+        if np.isscalar(self.sd):
+            self.sd = np.repeat(self.sd, self.size)
+        self.mean = np.asarray(self.mean)
+        self.sd = np.asarray(self.sd)
+        if self.mean.size != self.size:
+            raise ValueError("Mean vector size not matching.")
```
Comment (kels271828) [inline src/regmod/prior.py]: Suggestion: "Mean vector size does not match."

File: src/regmod/prior.py [hunk 5]
```diff
@@ -27,30 +66,137 @@ def process_size(self, vecs: List[Any]):
...
+        if np.isscalar(self.ub):
+            self.ub = np.repeat(self.ub, self.size)
+        self.lb = np.asarray(self.lb)
+        self.ub = np.asarray(self.ub)
+        if self.lb.size != self.size:
+            raise ValueError("Lower bound vector size not matching.")
+        if self.ub.size != self.size:
+            raise ValueError("Upper bound vector size not matching.")
+        if any(self.lb > self.ub):
+            ValueError("Lower bounds must be less or equal than upper bounds.")
```
Comment (kels271828) [inline src/regmod/prior.py]: Suggestion: "Lower bounds must be less than or equal to upper bounds."

File: src/regmod/prior.py [hunk 6]
```diff
@@ -62,11 +208,59 @@ def __post_init__(self):
...
+    size : int, default=100
+        Size of the spline prior. Default is 100. It determines the number
+        of sample points in the specified domain.
+    order : int, default=0
+        Order of the spline derivative. Default is 0.
+    domain_lb : float, default=0.0
+        Lower bounds of the domain. Default is 0.0.
+    domain_ub : float, default=1.0
+        Upper bounds of the domain. Default is 1.0.
+    domain_type : {'abs', 'rel'}, default='rel'
```
Comment (kels271828) [inline src/regmod/prior.py]: When the options are listed, I think the default typically comes first (see https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard, section 4)

**Comment by zhengp0** [comment]
Thanks very much @kels271828 ! I made all the changes, let me know if it is alright :)

**Review by kels271828 — APPROVED** [review]

**Pull request merged**